### PR TITLE
Extend MarkdownContent to allow ReactMarkdown to render trusted html …

### PR DESCRIPTION
**Pull request by mryave**
…content embedded in gfm

Often there is trusted HTML in Github flavored generated markdown.  This PR allows the use of the ReactMarkdown feature that allows trusted HTML to be processed. 

Before:
Screen Shot 2022-04-20 at 1 18 57 PM
After:
Screen Shot 2022-04-20 at 1 18 51 PM

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/core-components/src/components/MarkdownContent/MarkdownContent.tsx [hunk 1]
```diff
@@ -88,14 +90,17 @@ const components: Options['components'] = {
  * MarkdownContent
  * --
  * Renders markdown with the default dialect [gfm - GitHub flavored Markdown](https://github.github.com/gfm/) to backstage theme styled HTML.
+ * Pass html to activate [rehypeRaw](https://github.com/remarkjs/react-markdown#appendix-a-html-in-markdown) styling that will render any inlined HTML in gfm.
+ * demo of remark-gfm and rehype-raw working together and separate available [here](https://remarkjs.github.io/react-markdown/)
  * If you just want to render to plain [CommonMark](https://commonmark.org/), set the dialect to `'common-mark'`
  */
 export function MarkdownContent(props: Props) {
-  const { content, dialect = 'gfm', linkTarget } = props;
+  const { content, dialect = 'gfm', linkTarget, html = false } = props;
```
Comment (Rugvip) [inline packages/core-components/src/components/MarkdownContent/MarkdownContent.tsx:98]: Thoughts on calling it `trustHtml` or something like that? Just to make it clear that there's an additional level of trust that's needed. I'm assuming that this does not open up for script execution in any way right? But rather things like clickjacking through some creative styling.

To be clear as well this option is separate from the dialect right? so supports both common-mark and gfm

File: ADOPTERS.md [hunk 2]
```diff
@@ -113,5 +113,6 @@ _If you're using Backstage in your organization, please try to add your company
...
+| [Resuelve Tu Deuda](https://resuelve.mx)                                    | [Iván Álvarez](https://github.com/ivanhoe), [Jorge Medina](https://github.com/jorgearma1982) | Internal developer portal, service catalog, tech docs, api doc
+| [WeWork](https://www.wework.com)                                            | [David Szarzynski](https://github.com/Dithilli), [Mehera Pokale](https://github.com/mpokale), [Michael Ryave](https://github.com/mryave) | Internal developer experience & discovery portal. Team: Developer Experience
```
Comment (Rugvip) [inline ADOPTERS.md:118]: Wanna ship this through a separate PR rather than sneak it in here? :grin:

**Review by Rugvip — COMMENTED** [review]
Interesting! Like the idea but wanna be a bit careful. Feels pretty safe that it's opt-in though, but don't want it to be too much of a footgun.

A separate concern is bundle size, do you have a sense for how much this is adding? It does seem like it's adding a completely separate html parser

File: packages/core-components/src/components/MarkdownContent/MarkdownContent.tsx (hunk 1 above)
Comment (mryave) [reply packages/core-components/src/components/MarkdownContent/MarkdownContent.tsx:98]: Great points I also had to ponder and happy to talk more about. What did drive this implementation was the assumption of a trusted environment:
- totally agree to add something like `trustedHtml` to the prop name.  ReactMarkdown calls out that any HTML [would need to be in a trusted atmosphere to be totally safe](https://github.com/remarkjs/react-markdown#appendix-a-html-in-markdown).  Script execution shouldn't be able to happen, but clickjacking yes.  On that note, we can consider using rehype-sanitize. However, going that route does take away some features like being able to click ids and autofocus to parts of a page and doesn't exactly do what github.com actually supports, so the point of using it was defeated for the feature set this extension would try to accomplish.  
- This does support both dialects currently, which made this extension actually easier.

I'm happy to go with whatever you'd prefer.

**Comment by mryave** [comment]
> Interesting! Like the idea but wanna be a bit careful. Feels pretty safe that it's opt-in though, but don't want it to be too much of a footgun.
> 
absolutely, if this doesn't fit we'll have to keep this as a fork in our own repo. i do think this is something a lot of users will need at some point when it comes to Backstage for reading documentation. We generate markdown for all different kinds of documentation at wework. The first place I hope to use this feature directly though is for our gRPC documentation generated from [protoc-gen-doc](https://github.com/pseudomuto/protoc-gen-doc) which provides an excellent experience for reading and sharing gRPC documentation. I was originally using ReactMarkdown in a separate PR specific to the api-docs plugin until i found Backstage has its own core component that was doing the exact things i was doing. 

I have a follow up PR after this one to add MarkdownApiDocumentation as an option in @backstage/api-docs.
> A separate concern is bundle size, do you have a sense for how much this is adding? It does seem like it's adding a completely separate html parser

I wasn't sure if this would be called out.  it adds about 60kb, but for a feature set that i think is worth that weight.  This allows a richer Markdown experience for folks using Backstage. When it comes to generated markdown documentation, i feel html inside markdown is common. Once i discovered Backstage's MarkdownContent was using ReactMarkdown under the hood, I figured we might as well use its feature set rather than reinvent it elsewhere even though i feel @backstage/tech-docs might be flirting with that already.

Happy to do whatever works for the community here.

**Comment by jrusso1020** [comment]
For context we also use protobuf docs generated from protoc-gen-doc internally and we created our own custom markdown rendering component (taken from the react-markdown preview demo) https://github.com/uiwjs/react-markdown-preview/blob/master/src/index.tsx

I definitely would also see value in having a more robust markdown rendering component upstream, although our use case appears to be exactly the same (not a novel usecase)

**Comment by Rugvip** [comment]
One of my biggest worries here is actually test execution. I'm feeling it would be alright to add this if you can do it in a way that has 0 impact on tests unless they use the feature. What that means in practice I think would be to make sure this is all lazy loaded with `import`. That also addresses bundle size and page rendering speed too.

Is that something you could explore? Basically if `trustedHtml` is set then the component loads in `rehype-raw` and doesn't continue rendering until it is available, probably falling back to the app `Progress` component, or if we can have it inline in the markdown element that'd work too.

**Pull request closed without merging**